Releases/4 2 4

jacksonsj opened the pull request:
**IMPORTANT: Please do not create a Pull Request without creating an issue first.**

## Summary of Proposed Changes
<!--- Bulleted description of what this code changes, adds, removes, or resolves in the package
- Adds functionality to allow...
- Resolves bug associated with...
- Removes redundant code in...
--->

## Affected Process Elements
- [ ] Data Pack generation
- [ ] Data Pack model
- [ ] Data Pack self-service app
- [ ] Data Pack processing
- [ ] Site Tool model/distribution
- [ ] Site Tool generation (from Data Pack)
- [ ] Site Tool generation (from DATIM - for OPUs)
- [ ] Site Tool self-service app
- [ ] Site Tool to Data Pack comparison self-service app
- [ ] Site Tool to DATIM comparison self-service app
- [ ] Site Tool processing
- [ ] Site Tool import

## Types of changes

<!--- What types of changes does your code introduce to Appium? --->
<!--- _Put an `x` in the boxes that apply_ --->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## Does this introduce a breaking change?

- [ ] Yes
- [ ] No

<!-- If this introduces a breaking change, please describe the impact and migration path for existing applications below. -->

## Related Issues
<!--- Include list of issues resolved by this PR: --->
- Resolves #01
- Fixes #02
- Closes #03

<!--- The following keywords, followed by an issue number, will close the referenced issue:

- close
- closes
- closed
- fix
- fixes
- fixed
- resolve
- resolves
- resolved

To close an issue in another repository, use the `username/repository#issue_number` syntax
--->

## Testing Approach
<!--- How has this code been tested? Where helpful, include screenshots --->

## Final Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

sam-bao commented:
@jacksonsj I am getting failing unit tests on 4.2.4

```
> devtools::test()
Loading datapackr
Testing datapackr
✓ |  OK F W S | Context
✓ |   3       | Test GetDataValueSets [0.2 s]
x |   2 1 2   | test-get-datapack-name [16.8 s]                                                       
────────────────────────────────────────────────────────────────────────────────────────────────────────
test-get-datapack-name.R:22: error: Can read a Datapack Name and UIDs
Deduced or provided Country UIDs do no match Country UIDs observed in submission.
Backtrace:
 1. datapackr::unPackCountryUIDs(...) tests/testthat/test-get-datapack-name.R:22:2

test-get-datapack-name.R:38: warning: Can error on an invalid regional DataPack UID
Unable to deduce Country UIDs from cell B25 on Home tab. Attempting to deduce from org units listed on Prioritization tab instead.

test-get-datapack-name.R:38: warning: Can error on an invalid regional DataPack UID
Unable to deduce Country UIDs from Prioritization tab. Attempting to deduce from Data Pack name on Home tab.
────────────────────────────────────────────────────────────────────────────────────────────────────────
✓ |   2       | test-handshake
✓ |  18       | dedupe [0.4 s]
x |   0 2     | test-signature [0.5 s]                                                                
────────────────────────────────────────────────────────────────────────────────────────────────────────
test-signature.R:5: error: Can generate a signature
Deduced or provided Country UIDs do no match Country UIDs observed in submission.
Backtrace:
 1. datapackr:::createKeychainInfo(...) tests/testthat/test-signature.R:5:2
 2. datapackr::unPackCountryUIDs(...) R/createKeychainInfo.R:142:4

test-signature.R:34: error: Can get the type and COP year of tool of a COP20 Data Pack
Deduced or provided Country UIDs do no match Country UIDs observed in submission.
Backtrace:
 1. datapackr:::createKeychainInfo(submission_path = test_sheet("COP20_Data_Pack_Template_vFINAL.xlsx")) tests/testthat/test-signature.R:34:3
 2. datapackr::unPackCountryUIDs(...) R/createKeychainInfo.R:142:4
────────────────────────────────────────────────────────────────────────────────────────────────────────
✓ |   5       | Test flatten data pack model for cop 19 function

══ Results ═════════════════════════════════════════════════════════════════════════════════════════════
Duration: 18.0 s

OK:       30
Failed:   3
Warnings: 2
Skipped:  0
```

sam-bao commented:
differences between 4.2.4 and master d object

Datapack: With PSNUxIM/Data Pack_Malawi_20210121230425.xlsx
- 4.2.4 d$data$SUBNATT_IMPATT excludes 255 records that were on master all for PMTCT_STAT_SUBNAT.N.T
- 4.2.4 d$data$analytics has psnu_uid column and master has psnuid column
- “- Different types for column `fiscal_year`: character vs double\n” 4.2.4 is a number
- The 4.2.4 analyics object contains AGYW_PREV.D.T and AGYW_PREV.D.T while master doesn’t and master contains PMTCT subnat data but 4.2.4 doesn’t
- On master d$datim only contains MER and subnat_impatt, 4.2.4 has additional objects (I believe expected)
- D$datim$mer on 4.2.4 contains AGYW_PREV data that was not parsed using master.
- D$datim$subnat_impatt has quite a lot of differences between master and 4.2.4 I suggest we just take a close look at 4.2.4 and make sure it is correct.

jacksonsj commented:
Responding to the issues you flagged @sam-bao:

4.2.4 d$data$SUBNATT_IMPATT excludes 255 records that were on master all for PMTCT_STAT_SUBNAT.N.T
_As long as you're still seeing `PMTCT_STAT_SUBNAT.N.Known.Pos.T`, `PMTCT_STAT_SUBNAT.N.New.Pos.T`, and `PMTCT_STAT_SUBNAT.N.New.Neg.T`, this is ok. The total numerator `PMTCT_STAT_SUBNAT.N.T` is not a target we import, so it should be expected to see this dropped during extraction._

4.2.4 d$data$analytics has psnu_uid column and master has psnuid column
_Expected. Updated to integrate across `datapackr` and `datapackr-app`._

“- Different types for column fiscal_year: character vs double\n” 4.2.4 is a number
_This is ok and is the result of defining predecessor parameters for `FY` to be supplied by users as numbers, which is more intuitive for users than supplying FY as a string in quotes. All dependent functions have been updated to reflect this change._

The 4.2.4 analyics object contains AGYW_PREV.D.T and AGYW_PREV.D.T while master doesn’t and master contains PMTCT subnat data but 4.2.4 doesn’t
_Expected. I added code to port AGYW into this object. Before, this wasn't present because it isn't disaggregated by IM and so wasn't in the PSNUxIM object where all the other MER data comes from.

For PMTCT, I believe this is related to still needing changes to SUBNAT metadata to be pushed from `test-mer2` to DATIM prod, but if you can send specifics I can confirm._

On master d$datim only contains MER and subnat_impatt, 4.2.4 has additional objects (I believe expected)
_Expected. I split the subnat_impatt data up into dataset/FY chunks to allow selective importing. We're expecting that DATIM won't be ready to receive FY21 SUBNAT targets at the same time as the FY22 ones. And we also have places in `datapackr` where it's convenient to have easy access to just get the prioritizations, which are also separated here._

D$datim$mer on 4.2.4 contains AGYW_PREV data that was not parsed using master.
_See above. Same explanation as where this appears in the analytics object._

D$datim$subnat_impatt has quite a lot of differences between master and 4.2.4 I suggest we just take a close look at 4.2.4 and make sure it is correct.
_Can you send specifics?_
